polar-core/src/kb.rs

```diff
@@ -75,11 +81,247 @@ impl KnowledgeBase {
+            .rules
+            .entry(name.clone())
+            .or_insert_with(|| GenericRule::new(name, vec![]));
+        generic_rule.add_rule(Arc::new(rule));
+    }
+
+    pub fn validate_rules(&self) -> PolarResult<()> {
+        for (symbol, generic_rule) in self.rules.clone() {
+            let generic_template = self.rule_templates.get(&symbol);
+            if let Some(generic_template) = generic_template {
+                // If a template with the same name exists, then the parameters must match for each rule
+                for (_, r) in generic_rule.rules {
+                    // get templates with correct arity
+                    let templates = generic_template
+                        .rules
+                        .values()
+                        .filter(|t| return t.params.len() == r.params.len())
+                        .collect::<Vec<&Arc<Rule>>>();
+
+                    if templates.is_empty() {
+                        let err: PolarError = error::RuntimeError::FileLoading {
+                            msg: format!("Rule {}({:#?}) not allowed because arity did not match rule template.", r.name, r.params),
+                        }
+                        .into();
+                        return Err(err.set_context(None, Some(&r.body)));
+                    } else {
+                        // these templates should all have matching arity
+                        // TODO: check params of rule against rule templates
+                        let mut found_match = false;
+                        for t in templates {
+                            if self.rule_params_match(r.as_ref(), t.as_ref()) {
+                                found_match = true;
+                                break;
+                            }
+                        }
+                        if !found_match {
+                            let err: PolarError = error::RuntimeError::FileLoading {
+                                msg: format!("Rule {}({:#?}) not allowed because parameters did not match rule template.", r.name, r.params),
+                            }
+                            .into();
+                            return Err(err.set_context(None, Some(&r.body)));
+                        }
+                    }
+                }
+            }
+        }
+        Ok(())
+    }
+
+    fn param_fields_match(&self, template_fields: &Dictionary, rule_fields: &Dictionary) -> bool {
+        return template_fields
+            .fields
+            .iter()
+            .map(|(k, template_value)| {
+                rule_fields
+                    .fields
+                    .get(k)
+                    .map(|rule_value| rule_value == template_value)
+                    .unwrap_or_else(|| false)
+            })
+            .all(|v| v);
+    }
+
+    fn rule_params_match(&self, rule: &Rule, template: &Rule) -> bool {
+        for (rule_param, template_param) in rule.params.iter().zip(template.params.iter()) {
```

Comment:
You can use https://doc.rust-lang.org/std/iter/trait.Iterator.html#method.all here to clean up this code a bit. Breaking it into several functions may also help.

Reply:
Done.